[Daniel Kok] iP

danielk0k opened the pull request:
## THE ONLY DUKE YOU NEED

*Made with love :heart: for programmers by programmers.*

### Never miss another assignment again!

#### Features

- Execute commands by simply typing them like the command line
- Load in your tasks directly from your local file directory 
- Perform basic operations on your tasks like add, delete and mark as completed 
- Save your changes to your local file directory
- **And many more!**

#### How to use

1. Ensure that you have the correct version of Java installed on your machine
```zsh
$ java --version
```
> Supports only Java 11
2. Download `duke_v0.1.jar` file from [here](https://github.com/danielk0k/ip/releases/download/A-Jar/duke_v0.1.jar)
3. Double-click on the downloaded file and that's all!

#### More features to come

- [x] Search a task by typing `find taskKeyword`
- [ ] GUI interface using JavaFX

#### For the technical people

Run your own Duke program with the following main method in Java.
```java
public class Main {
    public static void main(String[] args) {
        Duke dk = new Duke("data/duke.txt");
        dk.run();
    }
}
```

Commit:
Add Gradle support

danielk0k committed:
Duke: Add Greet Echo Exit functions

danielk0k committed:
Duke: Add indentation and horizontal lines

danielk0k committed:
Duke: Add List and add functions

danielk0k committed:
Add new classes Task and TaskList

Duke: Add mark as done or not done function

danielk0k committed:
Add subclasses ToDo, Event, Deadline

Duke: Add support for tracking three types of tasks

danielk0k committed:
Duke: Add some exception handling

danielk0k committed:
Duke: Add delete function

danielk0k committed:
Test: Modified EXPECTED.TXT with the changes to Duke

danielk0k committed:
Duke: Improved exception handling

danielk0k committed:
Improved exception handling and message replies

danielk0k committed:
Add save and load from file feature to Duke

danielk0k committed:
Update date field of Task to LocalDate time

danielk0k committed:
Merge branch 'branch-Level-7'

danielk0k committed:
Merge branch 'branch-Level-8'

danielk0k committed:
Fix bug causing invalid inputs

danielk0k committed:
Refactor and organise code to greater OOP

danielk0k committed:
Group code into packages

danielk0k committed:
Add new commands and fix interaction between classes

danielk0k committed:
Organise classes into packages

danielk0k committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

danielk0k committed:
Add junit test cases

danielk0k committed:
Add find feature to search for a task by a keyword

Level-9. Find

danielk0k committed:
Add documentation to main classes in duke package

danielk0k committed:
Fix some coding standard issues

danielk0k committed:
Fix merge conflict in duplicate header comments

danielk0k committed:
Add header comments for class and methods in duke.task package

danielk0k committed:
Merge branch 'branch-A-CodingStandard'

danielk0k committed:
Add documentation to classes in duke.command package

danielk0k committed:
Merge branch 'branch-A-JavaDoc'

danielk0k committed:
Apply checkstyle and fix style violations

danielk0k committed:
Apply checkstyle and fix style violations

danielk0k committed:
Parser: fix style violation with header comment

danielk0k committed:
Update readme with message from PR

bokun2 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,45 @@
+ */
+public class Duke {
+    private final Storage storage;
+    private final TaskList taskList;
```
I am confused why this taskList should be final because it will be changed and reload, right? 🤔

danielk0k replied:
Understand, I think you're right here. It's also ambiguous to label it as final where it's not a constant.

bokun2 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,57 @@
+                fw.write(it.next().toStringSaveFormat());
+            }
+            fw.close();
+            Ui.showError("Successfully saved contents into duke.txt");
```
Perhaps can use another method in Ui? Using showError sounds not corresponding to successfully saving contents.

danielk0k replied:
Thanks for pointing it out, might have accidentally used the wrong method 😅 .

bokun2 commented on src/main/java/duke/command/MarkCommand.java:
```diff
@@ -0,0 +1,60 @@
+import duke.DukeException;
+import duke.Storage;
+import duke.TaskList;
+import duke.Ui;
```
I have noticed that in every class, you always keep the import statement in order! 👍

bokun2 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,99 @@
+            return new EmptyCommand();
+        } finally {
+            sc.close();
+        }
```
This part of code strictly follows the coding standard of switch-case part. 👍

bokun2 reviewed (commented):
The using of Enum and hash table is great. The code is clean and strictly follows the coding standard.

JonLamy commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,45 @@
+    private final Ui ui;
+
+    /**
+     * Returns an instance of Duke.
+     * @param filePath String location of duke.txt.
```
Small nitpick to leave a line between these 2 lines

JonLamy commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,99 @@
+        Scanner sc = new Scanner(input);
+        try {
+            String command = sc.next();
+            if (!CommandType.COMMAND_MAP.containsKey(command)) {
```
This conditional is a little hard to understand. Perhaps abstracting a method to make it more readable?

JonLamy commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,158 @@
+     * Used in writing tasks to file in Storage class.
+     * @return Iterator of the list of tasks.
+     */
+    public Iterator<Task> toSave() {
```
Method name is a little ambiguous

danielk0k replied:
Got it, have made the changes.

JonLamy commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,89 @@
+     * @param ui Ui object for displaying messages.
+     */
+    @Override
+    public void execute(Storage storage, TaskList taskList, Ui ui) {
+        switch (type) {
+        case TODO:
+            Ui.showMsg(taskList.add(new ToDo("0", description)));
+            break;
+        case DEADLINE:
+            Ui.showMsg(taskList.add(new Deadline("0", description, dateString)));
+            break;
+        case EVENT:
+            Ui.showMsg(taskList.add(new Event("0", description, dateString)));
+            break;
+        default:
+            return;
+        }
```
Nice and organizedd, easy to read!

danielk0k replied:
Thank you!

JonLamy reviewed (commented):
Nice packaging, code is organized and readable! Might be improved on with some minor re-naming here and there. Very cool 👍

danielk0k committed:
Add initial files and changes for JavaFX GUI

danielk0k committed:
Commands: cmd.execute now returns a string response

danielk0k committed:
Add WrongCommand to handle wrong command entered

danielk0k committed:
GUI: Move responses from the CLI to the GUI

danielk0k committed:
GUI: Fix truncation of response and scrolling

danielk0k committed:
Fix styling issues

danielk0k committed:
Update gradle setting

danielk0k committed:
Minor fix for messages

danielk0k committed:
Remove Ui class

danielk0k committed:
Clean up code with removal of Ui class

danielk0k committed:
Remove final keyword as it is ambiguous and not a constant

danielk0k committed:
TaskList: rename toSave to getIterator method

danielk0k committed:
JavaDocs: add empty line between description and param tag

danielk0k committed:
Parser: use guard clauses and clean code

danielk0k committed:
TaskList: add assertions to document assumptions

danielk0k committed:
Storage: change variable name filepath

danielk0k committed:
Merge pull request #1 from danielk0k/branch-A-CodeQuality

Merge Code Quality Branch

danielk0k committed:
Merge pull request #2 from danielk0k/branch-A-Assertions

Merge Assertions Branch

danielk0k committed:
MainWindow: edit welcome message

danielk0k committed:
Mark command: allow for mass marking of tasks

Extend the functionality of the mark command.

Users can mark multiple task at once instead of typing multiple mark commands.

Let's,
* modify mark command to take in an array of indexes
* edit parser to read in more than 1 integers from user input

danielk0k committed:
MarkCommand: improve execution of command using counter

danielk0k committed:
Unmark Command: allow for mass unmarking of tasks

Extend the functionality of the unmark command.

Users can mark multiple task as incomplete at once instead of typing multiple unmark commands.

Let's,
* modify unmark command to take in an array of indexes
* edit parser to read in more than 1 integers from user input

danielk0k committed:
Delete Command: allow for mass deleting of tasks

Extend the functionality of the delete command.

Users can delete multiple tasks at once instead of typing multiple delete commands.

Let's,
* modify delete command to take in an array of indexes
* edit parser to read in more than 1 integers from user input
* mark tasks to be deleted from array list

danielk0k committed:
Fix styling issues

danielk0k committed:
MarkCommand: edit header comment for equals method

danielk0k committed:
GUI: add profile icons with some styling

danielk0k committed:
GUI: Add username title and fix margins

danielk0k committed:
Change personality to spongebob from duke

danielk0k committed:
Code Quality: fix checkstyle errors

danielk0k committed:
MainWindow: add welcome msg

danielk0k committed:
ParserTest: fix test cases for mark, unmark and delete commands

danielk0k committed:
Add product screenshot

danielk0k committed:
Add user guide and edit readme

danielk0k committed:
Code Quality: fix header comments

danielk0k committed:
Storage: manage issue when spongebob.txt file is not found

danielk0k committed:
Update README.md
